### ISSUE TITLE
Local Explanations with TFT

### ISSUE THREAD

**Issue by pdurham2** [jdb78/pytorch-forecasting#932]
Hello, I have more of a question than an issue. I am trying to extract feature explanations for each prediction using shap values, but am running into challenges with the current support for pytorch models in shap since the deep/gradient explainers cannot handle embeddings. Do you have any examples or recommendations for how to get these explanations, either with shap or with other methods?

I also looked at the weights of the variable selection networks with model.interpret_output() but there seemed to be little variation in the values across different predictions.

Thank you!

**Comment by EsmaeilFarhang** [comment]
Hi, I'm having the same problem with shap and pytorch_forcasting. Darts package has one branch under development for explainability. They were trying to use support shap in darts package, but I was not able to use it.

https://github.com/unit8co/darts/pull/909

**Comment by sairamtvv** [comment]
I found across this
[https://captum.ai/tutorials/Bert_SQUAD_Interpret](url)

They are able to handle the embeddings. I hope this helps.

**Comment by sairamtvv** [comment]
![image](https://user-images.githubusercontent.com/50791255/204724905-5a8dbf55-437d-4a88-8a9a-43ca8f504354.png)

**Comment by sairamtvv** [comment]
@pdurham2  Any success on local interpretation using TFT model. If so, Please help me also

**Comment by pdurham2** [comment]
@sairamtvv I had a bit of success with Captum, but the integrated gradients did not seem to vary much over time for a given time series (but that may just be my model). The trick that I had to do was to overwrite the TFT's forward to accept a tuple instead of a dict since Captum was having issues with the dict input. I think you can also do this with a function now like they do with `predict` in https://captum.ai/tutorials/Bert_SQUAD_Interpret

```
class Interpretable_TFT(TemporalFusionTransformer):
  def __init__(self, model):
    super().__init__(**model.hparams)
    self.load_state_dict(model.state_dict(), strict=False)
    
  def forward(self, encoder_cat, encoder_cont, encoder_target,
              encoder_lengths, decoder_cat, decoder_cont,
              decoder_target, decoder_lengths, decoder_time_idx,
              groups, target_scale
             ) -> Dict[str, torch.Tensor]:

    x_dict = {"encoder_cat": encoder_cat, 
              "encoder_cont": encoder_cont,
              "encoder_target": encoder_target,
              "encoder_lengths": encoder_lengths, 
              "decoder_cat": decoder_cat, 
              "decoder_cont": decoder_cont,
              "decoder_target": decoder_target,
              "decoder_lengths": decoder_lengths,
              "decoder_time_idx": decoder_time_idx,
              "groups": groups, 
              "target_scale": target_scale
             }

    output = super().forward(x_dict)
    return output.prediction
```

**Comment by sairamtvv** [comment]
@pdurham2  I hope you have plotted the partial dependency plots.  Those are varying and explanation of how each variable is effecting your output can be obtained

    def partial_dependency(self, best_tft, val_dataloader, filter_item_id=None, lst_input_feat=None):
        """
        Takes as an input list of input_features, calculates their upper and lower bounds,
        extract their partial dependency and save their partial dependency plots
        """
        if lst_input_feat is None:
            lst_input_feat = [<list of clomuns whose  dependency u wish to observe>]


        for input_feature in lst_input_feat:
            if self.data[input_feature].dtype == "category":
                upper_bound_feature = self.data[input_feature].nunique()
                lower_bound_feature = 0
                feature_range = np.arange(lower_bound_feature, upper_bound_feature, 1)

            else:
                upper_bound_feature = self.data[input_feature].max()
                lower_bound_feature = self.data[input_feature].min()
                feature_range = np.linspace(lower_bound_feature, upper_bound_feature, 100)

            dependency = best_tft.predict_dependency(val_dataloader.dataset,
                                                     input_feature, feature_range,
                                                     show_progress_bar=True,
                                                     mode="dataframe")

            # plotting median and 25% and 75% percentile
            agg_dependency = dependency.groupby(input_feature).normalized_prediction.agg(
                median="median", q25=lambda x: x.quantile(0.25), q75=lambda x: x.quantile(0.75)
            )
            ax = agg_dependency.plot(y="median")
            ax.fill_between(agg_dependency.index, agg_dependency.q25, agg_dependency.q75, alpha=0.3)
            ax.figure.savefig(f"{self.images_dir}/{input_feature}_partial.png")



Hope this helps

**Comment by sairamtvv** [comment]
The partial dependency varies for each time series
